applications/ShallowWaterApplication/custom_elements/primitive_var_element.cpp

```diff
@@ -0,0 +1,447 @@
+
+        // Check that the element's nodes contain all required SolutionStepData and Degrees of freedom
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            Node<3> &rnode = this->GetGeometry()[i];
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(VELOCITY,rnode)
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(HEIGHT,rnode)
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(PROJECTED_VECTOR1,rnode)
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(PROJECTED_SCALAR1,rnode)
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(BATHYMETRY,rnode)
+            KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(RAIN,rnode)
+
+            KRATOS_CHECK_DOF_IN_NODE(VELOCITY_X,rnode)
+            KRATOS_CHECK_DOF_IN_NODE(VELOCITY_Y,rnode)
+            KRATOS_CHECK_DOF_IN_NODE(HEIGHT,rnode)
+        }
+
+        return ierr;
+
+        KRATOS_CATCH("")
+    }
+
+//----------------------------------------------------------------------
+
+    template< unsigned int TNumNodes >
+    void PrimitiveVarElement<TNumNodes>::EquationIdVector(EquationIdVectorType& rResult, ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+        
+        unsigned int element_size = TNumNodes*3;
+        if(rResult.size() != element_size)
+            rResult.resize(element_size,false);                         // False says not to preserve existing storage!!
+        
+        GeometryType& rGeom = GetGeometry();
+        int counter=0;
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            rResult[counter++] = rGeom[i].GetDof(VELOCITY_X).EquationId();
+            rResult[counter++] = rGeom[i].GetDof(VELOCITY_Y).EquationId();
+            rResult[counter++] = rGeom[i].GetDof(HEIGHT).EquationId();
+        }
+        
+        KRATOS_CATCH("")
+    }
+
+//----------------------------------------------------------------------
+
+    template< unsigned int TNumNodes >
+    void PrimitiveVarElement<TNumNodes>::GetDofList(DofsVectorType& rElementalDofList,ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+        
+        const unsigned int element_size = TNumNodes*3;
+        if(rElementalDofList.size() != element_size)
+            rElementalDofList.resize(element_size);
+        
+        GeometryType& rGeom = GetGeometry();
+        int counter=0;
+        for (unsigned int i = 0; i < TNumNodes; i++)
+        {
+            rElementalDofList[counter++] = rGeom[i].pGetDof(VELOCITY_X);
+            rElementalDofList[counter++] = rGeom[i].pGetDof(VELOCITY_Y);
+            rElementalDofList[counter++] = rGeom[i].pGetDof(HEIGHT);
+        }
+        
+        KRATOS_CATCH("")
+    }
+
+//----------------------------------------------------------------------
+
+    template< unsigned int TNumNodes >
+    void PrimitiveVarElement<TNumNodes>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+        
```

Comment:
Same as previous commentaries, I have seen this code before, thing about use a base class instead